core/src/main/scala/org/apache/spark/sql/execution/ExistingPlans.scala

```diff
@@ -129,11 +135,22 @@ private[sql] object PartitionedPhysicalScan {
         val (a, f) = scanBuilderArgs
         val baseTableRDD = table.buildRowBufferRDD(Array.empty,
           a.map(_.name).toArray, f.toArray, useResultSet = false)
+
+        def resolveCol(left: Attribute, right: AttributeReference) =
+          columnScan.sqlContext.sessionState.analyzer.resolver(left.name, right.name)
+
         val rowBufferScan = RowTableScan(output, baseTableRDD, numBuckets,
           Seq.empty, table)
-        val bufferExchange = ShuffleExchange(columnScan.outputPartitioning,
-          rowBufferScan)
-        ZipPartitionScan(columnScan, bufferExchange)
+        val otherPartKeys = partitionColumns.map(_.transform {
+          case a: AttributeReference => rowBufferScan.output.find(resolveCol(_, a)).getOrElse {
+            throw new AnalysisException(s"RowBuffer output column $a not found in " +
+                s"${rowBufferScan.output.mkString(",")}")
+          }
+        })
+        assert(columnScan.outputPartitioning.satisfies(
+          ClusteredDistribution(columnScan.partitionColumns)))
+        ZipPartitionScan(columnScan, columnScan.partitionColumns,
+          rowBufferScan, otherPartKeys)
```

Comment:
thanks to point out that RDD[InternalRow] can be wrapped with RowDataSourceScanExec and plugged into ZipPartitionScan , removing otherRDD in ZPS
@vivekwiz it will be something like this now whenever you intend to use.
```
        ZipPartitionScan(columnScan, columnScan.partitionColumns,
          RowDataSourceScanExec(
            columnScan.output, reservoirRDD, relation, columnScan
                .outputPartitioning,
            Map.empty, None
          ),
          columnScan.partitionColumns
        )


```